content/posts/2018-01-10-kotlin-null.md

```diff
@@ -0,0 +1,249 @@
+Yep, just deal with it :sunglasses:. So let’s ask the “How” question instead of the “Why” one.
+
+# Enter Kotlin
+
+The drug of choice of mine at this point in time is Kotlin,
+but you cannot forget about Java when using Kotlin.
+Well, the JVM probably can be avoided by using Kotlin Native,
+but at least on Android you are stuck with the JVM as an intermediate bytecode provider.
+
+Kotlin was, and is a great language.
+Backed by a team at JetBrains behind the best IDE platform available,
+reasonable and pragmatic design decisions — what else to wish for?
+And, more importantly, it brought the concept of Null Safety
+to the production table for developers.
+
+All Java objects are essentially implicitly can be `null`,
+when Kotlin objects have to be declared nullable explicitly,
+otherwise the type system assumes them be to be non-nullable by default.
+
+Finally! We can use the JVM and forget about the `NullPointerException`!
+[Well, not exactly](https://en.wikipedia.org/wiki/Forgot_About_Dre).
+
+Let’s take a look at a very short Kotlin + RxJava sample.
+
+```kotlin
+Observable.just(null).subscribe()
+```
+
+This code will not be highlighted by the IDE, compiler will not show any warnings or errors and yet it will crash _in runtime_.
+Well, at least when using RxJava 2.1.8.
+
+Why? Fortunately enough, [RxJava is open source](https://github.com/ReactiveX/RxJava), so we can take a look ourselves.
+
+```java
+public static <T> Observable<T> just(T item) {
+    ObjectHelper.requireNonNull(item, "The item is null");
+    return RxJavaPlugins.onAssembly(new ObservableJust<T>(item));
+}
+```
+
+In other words, the underlying code checks the `item` to not be `null`. And yet, Kotlin did not save us.
+Which reminds me, I probably should trademark a sign “Kotlin — False God”...
+
+Anyway, the reason is quite simple. When using RxJava we are using Java bytecode which does not have enough metadata.
+To actually have it the Java code should be annotated with so-called Nullability Annotations.
+The platform itself does not have enough introspection knowledge about if a value can be `null` or not.
+
+OK, so now we’ve established that we need annotations in the Java code to make Kotlin happy.
+In other words, the RxJava source code should be changed to be like this.
+
+```java
+public static <T> Observable<T> just(@NonNull T item) {
+    ObjectHelper.requireNonNull(item, "The item is null");
+    return RxJavaPlugins.onAssembly(new ObservableJust<T>(item));
+}
+```
+
+The difference is in the `@NonNull` annotation. The IDE will highlight the original code as an error,
+the Kotlin compiler will stop compilation... The End.
+
+It is totally another story about mutiple _different_ Nullability Annotations artifacts
+from JetBrains, FindBugs, The Checker Framework, Android Support Library and others,
+but [Jesse has it covered](https://medium.com/square-corner-blog/rolling-out-nullable-42dd823fbd89).
+
+# The Unicorn
+
```

Comment:
Little bit nitpicking, but in general you should avoid 
TITLE
SUBTITLE

without any text between it.

Also (it depends on your css style) but readers usually can't distinguish easily between TITLE / SUBTITLE  (or Chapter / Section / Subsection) just by looking at the style of the element. Professional books solve this problem by using numbers like  `1. Chapter A` and `1.1 Section X` and `1.1.1 Subsection Y` ... 

Again, this might be a little bit nitpicking and you don't have to agree with me,  I'm just saying ...

Reply:
Yeah, The Unicorn title feels out of place, I’m going to flatten things. Thanks, Hannes!